src/bit_machine/mod.rs

```diff
@@ -365,67 +369,74 @@ impl BitMachine {
-            return Ok(());
-        }
-        let mut src_buf = vec![0 as UWORD; a_frame_size + b_frame_size];
-        let src_ptr_end = unsafe { src_buf.as_mut_ptr().add(a_frame_size) }; // A frame write
-        let src_ptr = src_buf.as_mut_ptr(); // A read frame at ptr start
-        let dst_ptr_begin = unsafe { src_buf.as_mut_ptr().add(a_frame_size) }; // B read frame at ptr begin
-        let dst_ptr_end = unsafe { src_buf.as_mut_ptr().add(a_frame_size + b_frame_size) }; // B write frame at ptr end
-
-        // For jet from type A -> B
-        // Jets execution: There is single buffer with a_frame_size + b_frame_size UWORDs
-        // ------[ A read frame     ][    B write frame  ]---
-        //       ^ src_ptr         ^src_ptr_end(dst_ptr_begin)      ^ dst_ptr_end
-        // 1. Write into C bitmachine using A write frame(= src_ptr_end)
-        // Precondition satisfied: src_ptr_end is one past the end of slice of UWORDs for A.
-        let mut a_frame = unsafe { CFrameItem::new_write(src_ty_bit_width, src_ptr_end) };
-        for _ in 0..src_ty_bit_width {
+        // Write bits from active read frame into input frame
+        let buffer_end = unsafe { buffer.as_mut_ptr().add(buffer_width) };
+        let mut write_frame = unsafe { CFrameItem::new_write(bit_width, buffer_end) };
+        for _ in 0..bit_width {
             let bit = self.read_bit();
             unsafe {
-                c_writeBit(&mut a_frame, bit);
+                c_writeBit(&mut write_frame, bit);
             }
         }
-        self.back(src_ty_bit_width);
+        self.back(bit_width);
 
-        // 2. Execute the jet. src = A read frame, dst = B write frame
-        // Precondition satisfied: src_ptr is the start of slice of UWORDs of A.
-        let src_frame = unsafe { CFrameItem::new_read(src_ty_bit_width, src_ptr) };
-        // Precondition satisfied: dst_ptr_end is one past the end of slice of UWORDs of B.
-        let mut dst_frame = unsafe { CFrameItem::new_write(target_ty_bit_width, dst_ptr_end) };
-        let jet_fn = jet.c_jet_ptr();
-        let c_env = jet.c_jet_env(env);
-        let res = jet_fn(&mut dst_frame, src_frame, c_env);
+        // Return input frame as read frame
+        let buffer_ptr = buffer.as_mut_ptr();
+        unsafe { CFrameItem::new_read(bit_width, buffer_ptr) }
+    }
+
+    /// Return a write frame that is equivalent to the active write frame.
+    fn get_output_frame(bit_width: usize, buffer: &mut Vec<UWORD>) -> CFrameItem {
+        let buffer_width = uword_width(bit_width);
+        buffer.clear();
+        buffer.resize(buffer_width, 0);
 
-        if !res {
+        // Return output frame as write frame
+        let buffer_end = unsafe { buffer.as_mut_ptr().add(buffer_width) };
+        unsafe { CFrameItem::new_write(bit_width, buffer_end) }
+    }
+
+    /// Return the value of the given type that is contained in the buffer.
+    fn read_output_frame(&mut self, bit_width: usize, buffer: &mut Vec<UWORD>) {
```

Comment:
In 63c2d233fb29a8837e930d79507d89d06a4443d9:

Here you can take a `&[UWORD]` and you don't even need `mut`.